spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaConsumerFactory.java

```diff
@@ -202,9 +242,49 @@ private void checkForUnsupportedProps(Properties properties) {
 		});
 	}
 
-	protected KafkaConsumer<K, V> createKafkaConsumer(Map<String, Object> configProps) {
-		return new KafkaConsumer<>(configProps, this.keyDeserializerSupplier.get(),
+	@SuppressWarnings("resource")
+	protected Consumer<K, V> createKafkaConsumer(Map<String, Object> configProps) {
+		Consumer<K, V> kafkaConsumer = createRawConsumer(configProps);
+
+		if (this.listener != null) {
+			Map<MetricName, ? extends Metric> metrics = kafkaConsumer.metrics();
+			Iterator<MetricName> metricIterator = metrics.keySet().iterator();
+			String clientId;
+			if (metricIterator.hasNext()) {
+				clientId = metricIterator.next().tags().get("client-id");
+			}
+			else {
+				clientId = "unknown";
+			}
+			String id = this.beanName + "." + clientId;
+			kafkaConsumer = createProxy(kafkaConsumer, id);
+			this.listener.consumerAdded(id, kafkaConsumer);
+		}
+		return kafkaConsumer;
+	}
+
+	protected Consumer<K, V> createRawConsumer(Map<String, Object> configProps) {
+		Consumer<K, V> kafkaConsumer = new KafkaConsumer<>(configProps, this.keyDeserializerSupplier.get(),
 				this.valueDeserializerSupplier.get());
+		return kafkaConsumer;
+	}
+
+	@SuppressWarnings("unchecked")
+	private Consumer<K, V> createProxy(Consumer<K, V> kafkaConsumer, String id) {
+		ProxyFactory pf = new ProxyFactory(kafkaConsumer);
+		Advice advice = new MethodInterceptor() {
+
+			@Override
+			public Object invoke(MethodInvocation invocation) throws Throwable {
+				DefaultKafkaConsumerFactory.this.listener.consumerRemoved(id, kafkaConsumer);
+				return invocation.proceed();
+			}
+
+		};
+		NameMatchMethodPointcutAdvisor advisor = new NameMatchMethodPointcutAdvisor(advice);
+		advisor.addMethodName("close");
```

Comment:
I think creating a proxy just for a `close()` method is an overhead. 
It probably might be more performant is we would just return a wrapper.

Although probably similar. So, just my observation.

Reply:
The problem with wrappers is you have to use reflection when the underlying API changes (new methods added etc).

I doubt the overhead is more than noise; but I'll run some tests.